Index: src/hotspot/cpu/riscv/frame_riscv.hpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/hotspot/cpu/riscv/frame_riscv.hpp	(revision 8afdcc7532a76dda0c2318b7ece7d18ff82aa33a)
+++ src/hotspot/cpu/riscv/frame_riscv.hpp	(date 1587401395849)
@@ -23,6 +23,9 @@
  *
  */
 
+#ifdef CMAKE_DEF
+#undef CPU_RISCV_FRAME_RISCV_HPP
+#endif
 #ifndef CPU_RISCV_FRAME_RISCV_HPP
 #define CPU_RISCV_FRAME_RISCV_HPP
 
Index: src/hotspot/os/linux/os_linux.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/hotspot/os/linux/os_linux.cpp	(revision 8afdcc7532a76dda0c2318b7ece7d18ff82aa33a)
+++ src/hotspot/os/linux/os_linux.cpp	(date 1587408516978)
@@ -1824,7 +1824,7 @@
   static  Elf32_Half running_arch_code=EM_68K;
 #elif  (defined SH)
   static  Elf32_Half running_arch_code=EM_SH;
-#elif  (defined __riscv)
+#elif  (defined __riscv) || (defined RISCV)
   static  Elf32_Half running_arch_code=EM_RISCV;
 #else
     #error Method os::dll_load requires that one of following is defined:\
Index: src/hotspot/share/runtime/sharedRuntime.cpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/hotspot/share/runtime/sharedRuntime.cpp	(revision 8afdcc7532a76dda0c2318b7ece7d18ff82aa33a)
+++ src/hotspot/share/runtime/sharedRuntime.cpp	(date 1587408582486)
@@ -383,7 +383,7 @@
 
 #endif
 
-#if defined(__SOFTFP__) || defined(PPC) || defined(__riscv)
+#if defined(__SOFTFP__) || defined(PPC) || defined(__riscv) || defined(RISCV)
 double SharedRuntime::dsqrt(double f) {
   return sqrt(f);
 }
Index: src/hotspot/share/utilities/macros.hpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/hotspot/share/utilities/macros.hpp	(revision 8afdcc7532a76dda0c2318b7ece7d18ff82aa33a)
+++ src/hotspot/share/utilities/macros.hpp	(date 1587408808488)
@@ -25,6 +25,28 @@
 #ifndef SHARE_UTILITIES_MACROS_HPP
 #define SHARE_UTILITIES_MACROS_HPP
 
+#ifdef CMAKE_DEF
+
+#define RISCV
+#define ASSERT
+#define RISCV64
+#define LINUX
+//#define __riscv
+
+#define offset_of offsetof
+
+#ifndef INCLUDE_SUFFIX_CPU
+#define INCLUDE_SUFFIX_CPU _riscv
+#define INCLUDE_SUFFIX_OS _linux
+#define INCLUDE_SUFFIX_COMPILER _gcc
+#endif
+#endif
+
+#ifndef intptr_t
+#include "stdint.h"
+
+#endif
+
 // Use this to mark code that needs to be cleaned up (for development only)
 #define NEEDS_CLEANUP
 
Index: src/hotspot/share/runtime/sharedRuntime.hpp
IDEA additional info:
Subsystem: com.intellij.openapi.diff.impl.patch.CharsetEP
<+>UTF-8
===================================================================
--- src/hotspot/share/runtime/sharedRuntime.hpp	(revision 8afdcc7532a76dda0c2318b7ece7d18ff82aa33a)
+++ src/hotspot/share/runtime/sharedRuntime.hpp	(date 1587408637907)
@@ -150,7 +150,7 @@
   static double dabs(double f);
 #endif
 
-#if defined(__SOFTFP__) || defined(PPC) || defined(__riscv)
+#if defined(__SOFTFP__) || defined(PPC) || defined(__riscv) || defined(RISCV)
   static double dsqrt(double f);
 #endif
 
